File: Programmers/level2/양궁대회.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;
vector<int> lion;
int answerSum;
void dfs(int k, int n, vector<int>& answer, vector<int>& info) {
    if (k == 11) {
        int aSum = 0, lSum = 0;
        for (int i = 0; i < 11; i++) {
            if (info[i] == 0 && lion[i] == 0)    continue;
            if (info[i] >= lion[i])    aSum += 10 - i;
            else lSum += 10 - i;
        }

        if (aSum < lSum && answerSum <= (lSum - aSum)) {
            if (answerSum == (lSum - aSum)) {
                for (int i = 10; i >= 0; i--) {
                    if (answer[i] > lion[i])   break;
                    else if (answer[i] < lion[i]) {
                        answer = lion;
                        break;
                    }
                }
            }
            else answer = lion;
            answerSum = lSum - aSum;
        }
        return;
    }
    for (int i = 0; i <= n; i++) {
        lion.push_back(i);
        dfs(k + 1, n - i, answer, info);
        lion.pop_back();
    }
}
vector<int> solution(int n, vector<int> info) {
    vector<int> answer(11, -1);
    answerSum = 0;
    dfs(0, n, answer, info);
    if (answer[0] == -1)   answer.assign(1, -1);
    return answer;
}

File: Programmers/level2/양궁대회.java
class Solution {
    static int maxDiff = 0;
    static int[] maxRyan = new int[11];

    static void dfs(int depth,int n,int[] apeach,int[] ryan){
        if(depth==n){
            int apeachSum=0;
            int ryanSum=0;
            for(int i=0;i<11;i++){
                if(apeach[i]==0 && ryan[i]==0){
                    continue;
                }
                if(apeach[i]>=ryan[i]){
                    apeachSum+=10-i;
                }else{
                    ryanSum+=10-i;
                }
            }
            if(ryanSum<=apeachSum)  return;

            if(ryanSum - apeachSum > maxDiff){
                maxDiff = ryanSum - apeachSum;
                maxRyan = ryan.clone();
            }
            else if(ryanSum - apeachSum == maxDiff){
                for(int i=10;i>=0;i--){
                    if(maxRyan[i]<ryan[i]){
                        maxRyan = ryan.clone();
                        break;
                    }else if(maxRyan[i]>ryan[i]){
                        break;
                    }
                }
            }

            return;
        }
        for(int i=0;i<=10;i++){
            if(ryan[i]>apeach[i]) break;
            ryan[i]+=1;
            dfs(depth+1,n,apeach,ryan);
            ryan[i]-=1;
        }
    }
    public int[] solution(int n, int[] info) {
        int[] answer = {};
        int[] ryan = new int[11];
        dfs(0,n,info,ryan);

        int sum=0;
        for(int i=0;i<11;i++){
            sum+=maxRyan[i];
        }
        if(sum==0){
            answer=new int[]{-1};
        }else{
            answer=maxRyan.clone();
        }
        return answer;
    }
}